internal/condition/condition.go

```diff
@@ -0,0 +1,306 @@
+package condition
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common"
+	celtypes "github.com/google/cel-go/common/types"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/openfga/openfga/internal/condition/types"
+	"github.com/openfga/openfga/internal/errors"
+	"golang.org/x/exp/maps"
+)
+
+var celBaseEnv *cel.Env
+
+func init() {
+	var envOpts []cel.EnvOption
+	for _, customTypeOpts := range types.CustomParamTypes {
+		envOpts = append(envOpts, customTypeOpts...)
+	}
+
+	envOpts = append(envOpts, types.IPAddressEnvOption(), cel.EagerlyValidateDeclarations(true))
+
+	env, err := cel.NewEnv(envOpts...)
+	if err != nil {
+		panic(fmt.Sprintf("failed to construct CEL base env: %v", err))
+	}
+
+	celBaseEnv = env
+}
+
+var emptyEvaluationResult = EvaluationResult{}
+
+type EvaluationResult struct {
+	Cost              uint64
+	ConditionMet      bool
+	MissingParameters []string
+}
+
+// EvaluableCondition represents a condition that can eventually be evaluated
+// given a CEL expression and a set of parameters. Calling .Evaluate() will
+// optionally call .Compile() which validates and compiles the expression and
+// parameter type definitions if it hasn't been done already.
+// Note: at the moment, this is not safe for concurrent use.
```

Comment:
Can we put this comment above the methods that aren't thread safe?

Reply:
removing this outdated bit, this is now safe for concurrent use.